SwanProjects/src/ProjectWidget.tsx

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) SWAN Development Team.
+// Author: Omar.Zapata@cern.ch 2021
+
+import * as React from 'react';
+
+import Button from '@material-ui/core/Button';
+import TextField from '@material-ui/core/TextField';
+import MenuItem from '@material-ui/core/MenuItem';
+
+import { Card, HelpTooltip } from './Components';
+export interface IStackOptions {
+  visible: boolean;
+}
+import { swanProjectIcon, sftIcon, cmsIcon } from './icons';
+
+import { ProjectDialog } from './ProjectDialog';
+
+export const ProjectWidget: React.FunctionComponent<{
+  options: ProjectDialog.ISWANOptions;
+  onSubmit: (selectedOptions: ProjectDialog.ISWANOptions) => void;
+  onCancel: () => void;
+}> = props => {
+  const options = props.options;
+  const [projectName, setProjectName] = React.useState(options.name || '');
+
+  const availableStacks = Object.keys(options.stacks_options);
+  const defaultStack = availableStacks.includes(options.stack)
+    ? options.stack
+    : availableStacks[0];
+  const [stack, setStack] = React.useState(defaultStack);
+
+  const availableReleases = Object.keys(options.stacks_options[stack]);
+  const defaultRelease = availableReleases.includes(options.release)
+    ? options.release
+    : availableReleases[0];
+  const [release, setRelease] = React.useState(defaultRelease);
+
+  const availablePlatforms = options.stacks_options[stack][release];
+  const defaultPlatform = availablePlatforms.includes(options.platform)
+    ? options.platform
+    : availablePlatforms[0];
+  const [platform, setPlatform] = React.useState(defaultPlatform);
+
+  const [userScript, setUserScript] = React.useState(options.user_script || '');
+
+  const onClickSubmit = () => {
+    props.onSubmit({
+      name: projectName,
+      stack,
+      release,
+      platform,
+      user_script: userScript,
+      stacks_options: options.stacks_options // TODO remove this
+    });
+  };
+
+  const onClickCancel = () => {
+    props.onCancel();
+  };
+
+  const onChangeProjectName = (event: React.ChangeEvent<HTMLInputElement>) => {
```

Comment:
N00b answer probably :) ,This was the way recommended here: https://reactjs.org/docs/forms.html of using controlled components. The uncontrolled approach mentioned looked harder at the time in react, involving using refs and looking up the ref on form submit. (https://goshakkk.name/controlled-vs-uncontrolled-inputs-react/)

One optimisation we can do here is to use React.useCallback/useMemo, to prevent the callback reference invalidating the props and trigger a re-render of sibling inputs on change. Although it may be a premature, as inputs are not expensive to render.

Reply:
This would add an extra step for example when changing the LCG release, we must edit platform options available. The reactive approach, the UI is a reflection of the state, and updates automatically/re-renders when the state changes, but everything lives in the state than the DOM.

But I do agree with you on the good ol'days....where life was simple